File: 2022/python/day03/solve.py
input = open("input.txt").read()
invs = input.splitlines()


def priority(item: chr):
    if ord('a') <= ord(item) <= ord('z'):
        return ord(item) - ord('a') + 1
    return ord(item) - ord('A') + 27


compartments = [[inv[0:int(len(inv)/2)], inv[int(len(inv)/2):]]
                for inv in invs]
common_items = [list(set(comp1).intersection(set(comp2)))[0]
                for [comp1, comp2] in compartments]
part1_sum = sum([priority(item) for item in common_items])


def chunks(lst, n):
    for i in range(0, len(lst), n):
        yield lst[i:i + n]


elf_groups = chunks(invs, 3)
common_items = [list(set(inv1).intersection(set(inv2)).intersection(set(inv3)))[0]
                for [inv1, inv2, inv3] in elf_groups]
part2_sum = sum([priority(item) for item in common_items])

print("Part 1:", part1_sum)
print("Part 2:", part2_sum)

File: 2022/haskell/day03/solve.hs
{-# LANGUAGE OverloadedStrings #-}

import Data.Char (ord)
import Data.List (intersect, sort)
import Data.List.Split (chunksOf)

parseInput :: String -> [[Int]]
parseInput input = [[priority c | c <- line] | line <- lines input]

priority :: Char -> Int
priority c
  | c >= 'a' = ord c - ord 'a' + 1
  | otherwise = ord c - ord 'A' + 27

halve :: [a] -> ([a], [a])
halve xs = splitAt (div (length xs) 2) xs

intersect' :: Eq a => ([a], [a]) -> [a]
intersect' (a, b) = a `intersect` b

solvePart1 :: String -> Int
solvePart1 = sum . map (head . intersect' . halve) . parseInput

solvePart2 :: String -> Int
solvePart2 = sum . map intersectGroup . chunksOf 3 . parseInput

intersectGroup :: [[Int]] -> Int
intersectGroup (inv1 : inv2 : inv3 : _) = head $ intersect (inv1 `intersect` inv2) inv3
intersectGroup _ = error "uh oh"

main :: IO ()
main =
  do
    input <- readFile "input.txt"
    putStrLn $ "Part one: " ++ show (solvePart1 input)
    putStrLn $ "Part two: " ++ show (solvePart2 input)
